### PR TITLE
Issue 19058 - Sort user-mentions suggestions by prefix first, then alphabetically

### LINKED ISSUE #19058: Mentions: Suggestions ordered by prefix instead of alphabetically.
Registering here @RenanLukas's feedback/suggestion while beta testing #18979 for feature enhancement.

> I have a question and maybe a suggestion about how the suggested user list is sorted. I’m sure the team discussed this, but I was thinking if it would be better to have the suggestions list sorted by prefix first (like “hasPrefix”) instead of alphabetically. Out of curiosity, I was testing other products and it seems like Slack checks the prefix and GitHub does it the way it is right now (alphabetically).
> I was just thinking that if the name prefix is common, there will be some scrolling before the user can find whoever they’re looking.
> 
> e.g.
> 
> Let’s say I start typing “Ren” to search for myself. The results are sorted alphabetically and apparently we check if the result item contain the searched string, so I was the 35th on the list when I did this test. If we had it sorted by the starting string (and alphabetically after that), it would look like:
> 1 – Renan
> 2 – <another name starting with Ren>
> 3 – <another name starting with Ren>
> …
> (and when we’re out of names that have “Ren” prefix, we’d be back to sorting alphabetically names that contain “Ren”)
> 4 – AAAAAren
> 5 – BBBBrenBB

### PR DESCRIPTION
**Pull request by igoriols**
Closes https://github.com/wordpress-mobile/WordPress-iOS/issues/19058

## Description

In this PR, I'm addressing the sorting change requested. I'm displaying first the users whose names begin with a prefix provided by the user, then alphabetically. The first criteria are to sort alphabetically by users whose usernames or display names begin with the provided prefix. If both username and display name begin with the prefix, the display name will be prioritized in the sorting. 

## Details

- I’m using localizedCaseInsensitiveCompare to compare `displayNames`, because the user can have the displayName in different languages. The comparison is localizable and case-insensitive.

- I’m using `<` to compare usernames because they are very restricted, and the lexicographical comparison should work for them.

- I'm looping through the `UserSuggestions` array to create two sub-arrays. The first one has the users whose display names begin with a prefix provided by the user. The second one has the users whose display names don't begin with the prefix provided by the user. 

- I'm sorting both arrays alphabetically, then I combine the sorted arrays. 

- I updated the tests to conform with the new sorting criteria.

## Preview

<p float="left">
  
</p>

## Testing Instructions

**N.B: The user-mentions suggestions feature is not enabled in self-hosted sites ( wordpress.org ).**

1. Tap the "My Site" tab.
2. Tap on "Menu"
3. Tap on "Comments" row.
4. Tap any comment.
5. Tap "Reply" button.
6. Type "@" and you should see the user-mentions suggestions list.
7. Type a prefix and check if the list is sorted first by users whose usernames or display names begin with the prefix, then alphabetically.

There are other scenarios to display the user-mentions suggestions list described in https://github.com/wordpress-mobile/WordPress-iOS/pull/18979.

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

9. What automated tests I added (or what prevented me from doing so)
I updated the existing tests related to the user-mentions suggestions list.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/Models/UserSuggestion+Comparable.swift [hunk 1]
```diff
@@ -0,0 +1,7 @@
+extension UserSuggestion: Comparable {
+    public static func < (lus: UserSuggestion, rus: UserSuggestion) -> Bool {
```
Comment (hassaanelgarem) [inline WordPress/Classes/Models/UserSuggestion+Comparable.swift]: [Nit]
```suggestion
    public static func < (lhs: UserSuggestion, rhs: UserSuggestion) -> Bool {
```

Comment (hassaanelgarem) [reply WordPress/Classes/Models/UserSuggestion+Comparable.swift]: I'm curious, what do `lus` and `ldn` stand for?

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift [hunk 2]
```diff
@@ -123,15 +123,35 @@ import CoreData
...
         }
     }
 
+    private static func sort(userSuggestions: [UserSuggestion], by searchQuery: String) -> [UserSuggestion] {
+        guard !searchQuery.isEmpty, !userSuggestions.isEmpty else { return userSuggestions }
+        var prefixedUserSuggestions = [UserSuggestion]()
+        var otherUserSuggestions = [UserSuggestion]()
+
+        userSuggestions.forEach { suggestion in
+            guard let displayName = suggestion.displayName else { return }
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]: This is a corner case, but if the user doesn't have a display name for some reason, they will be filtered out completely. I think we should add them to `otherUserSuggestions` in that case. Wdyt?

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift [hunk 3]
```diff
@@ -123,15 +123,35 @@ import CoreData
...
-                userSuggestions: userSuggestions,
+                userSuggestions: sortedUserSuggestions,
                 prominentSuggestionsIds: prominentSuggestionsIds
             )
         case .sites(let siteSuggestions):
             return .sites(siteSuggestions)
         }
     }
 
+    private static func sort(userSuggestions: [UserSuggestion], by searchQuery: String) -> [UserSuggestion] {
```
Comment (hassaanelgarem) [inline WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]: Can you add a brief documentation of what's happening here? 🙏

**Review by hassaanelgarem — COMMENTED** [review]
I think we can do two things differently:
- Doing a case insensitive search instead of case-sensitive
- Look for the prefix in both the username and the display name 

@igoriols What do you think?

File: WordPress/Classes/Models/UserSuggestion+Comparable.swift (hunk 1 above)
Comment (igoriols) [reply WordPress/Classes/Models/UserSuggestion+Comparable.swift]: lus: left user suggestion
rus: right user suggestion
ldn: left display name
rdn: right display name
I thought it would be better than the standard `lhs (left-hand side)` and `rhs (right-hand side)`, but I can change to use lhs and rhs. I can apply more meaningful names to the unwrapped display names, also.

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift (hunk 2 above)
Comment (igoriols) [reply WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]: Yes, that is a good idea. I'm going to apply it.

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift (hunk 3 above)
Comment (igoriols) [reply WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]: Sure. It's done. Thank you for reminding me.

**Comment by igoriols** [comment]
> I think we can do two things differently:
> 
>     * Doing a case insensitive search instead of case-sensitive
> 
>     * Look for the prefix in both the username and the display name
> 
> 
> @igoriols What do you think?

Hi @hassaanelgarem. Just to make sure I understand.

1. About case insensitive search, do you mean remove the `[c]` from [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift#L199)?

2. If the user has the prefix in `username` and `displayName`, should we sort by `displayName` first?
I'm asking because, in my opinion, the `displayName` is more relevant.

**Comment by hassaanelgarem** [comment]
> About case insensitive search, do you mean remove the [c] from [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift#L199)?

@igoriols No, I mean changing the `hasPrefix()` check [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/b43a6269f1ebd25477931de160c54d9f15b28df9/WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift#L147) to be case insensitive.

File: WordPress/WordPressShareExtension/String+Extensions.swift [hunk 4]
```diff
@@ -88,6 +88,13 @@ extension String {
         return self + returnURLString
     }
 
+    /// Returns a Boolean value indicating if this String begins with provided prefix. The comparison is both case-insensitive and diacritic-insensitive.
+    /// - Parameter prefix: The provideded prefix to verify.
+    /// - Returns: true if this string begins with the provided prefix, false otherwise.
+    func beginsWith(prefix: String) -> Bool {
+        return range(of: prefix, options: [.anchored, .caseInsensitive, .diacriticInsensitive]) != nil
+    }
+
```
Comment (AliSoftware) [inline WordPress/WordPressShareExtension/String+Extensions.swift]: Given this specifically does a case- and diacritics-insensitive test, which is what makes it differ from Foundation's built-in `String.hasPrefix`, I'd suggest naming this method accordingly, to avoid other devs starting to see "oh, autocomplete suggests me a `beginsWith(prefix:)` method on `String`, I'll use it for my feature" without realizing it's ignoring case and diacritics.

 - Maybe `hasCaseAndDiacriticsInsensitivePrefix` (though this is a bit of a mouthful)?
 - Or `hasFoldedPrefix` (with "folded" being used here in the sense if "String folding" as in [here](https://developer.apple.com/documentation/foundation/nsstring/1413779-folding))… though this term of art might not be widely known so maybe not super clear either…
 - Feel free to suggest anything else. I'm not super inspired (Naming is hard!) but maybe you have other ideas?

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift [hunk 5]
```diff
@@ -123,15 +123,42 @@ import CoreData
...
+        var prefixedUserSuggestions = [UserSuggestion]()
+        var otherUserSuggestions = [UserSuggestion]()
+
+        for suggestion in userSuggestions {
+            guard let username = suggestion.username else {
+                continue
+            }
+            if let displayName = suggestion.displayName,
+               (username.beginsWith(prefix: prefix) || displayName.beginsWith(prefix: prefix)) {
+                prefixedUserSuggestions.append(suggestion)
```
Comment (AliSoftware) [inline WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]: This will skip users with a `nil` `displayName`… even if they have a `username` that begins with the prefix.

So `{"ID":102,"user_login":"carlosdiaz","display_name":null}` will not be found even if `prefix` is `carl` here, because `let displayName = suggestion.displayName` will be false and exit the condition already.

Is that intentional? Do we want to prioritize only the users who have a non-nil display name?

If not, this condition will need to be adjusted accordingly 🙂  (and it would be nice to use the occasion to add a test case for this while we're at it 😉 )

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift [hunk 6]
```diff
@@ -123,15 +123,42 @@ import CoreData
...
+            } else {
+                otherUserSuggestions.append(suggestion)
+            }
+        }
+
+        return prefixedUserSuggestions.sorted() + otherUserSuggestions.sorted()
+    }
```
Comment (AliSoftware) [inline WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]: nitpick: May I suggest a simpler and more compact implementation, using `Dictionary(grouping:by:)`?

```suggestion
    private static func sort(userSuggestions: [UserSuggestion], by prefix: String) -> [UserSuggestion] {
        guard !userSuggestions.isEmpty, !prefix.isEmpty else { return userSuggestions }
        let triagedList = Dictionary(grouping: userSuggestions) { suggestion in
            suggestion.username?.beginsWith(prefix: prefix) == true || suggestion.displayName?.beginsWith(prefix: prefix) == true
        }
        return (triagedList[true] ?? []).sorted() + (triagedList[false] ?? []).sorted()
    }
```

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift [hunk 7]
```diff
@@ -123,15 +123,42 @@ import CoreData
...
-                userSuggestions: userSuggestions,
+                userSuggestions: sortedUserSuggestions,
                 prominentSuggestionsIds: prominentSuggestionsIds
             )
         case .sites(let siteSuggestions):
             return .sites(siteSuggestions)
         }
     }
 
+    /// Sort user suggestions by prefix first, then alphabetically. The collection is sorted first by checking if displayName begins with the provided prefix. The remaining items are sorted alphabetically by displayName. The prefix comparison is both case-insensitive and diacritic-insensitive.
```
Comment (AliSoftware) [inline WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]:
```suggestion
    /// Sort user suggestions by prefix first, then alphabetically. The collection is sorted first by checking if username or displayName begins with the provided prefix. The remaining items are sorted alphabetically by displayName. The prefix comparison is both case-insensitive and diacritic-insensitive.
```

File: WordPress/WordPressShareExtension/String+Extensions.swift (hunk 4 above)
Comment (igoriols) [reply WordPress/WordPressShareExtension/String+Extensions.swift]: Hi @AliSoftware , I see your point. I don't like a big name for this function, either. What if we keep the name `hasPrefix` and expose the options as a parameter allowing the user to define the compare options?

```suggestion
    /// Returns a Boolean value indicating if this String begins with provided prefix.
    /// - Parameter prefix: The provided prefix to verify.
    /// - Parameter options: The provided compare options.
    /// - Returns: true if this String begins with the provided prefix following the given compare options, false otherwise.
    func hasPrefix(_ prefix: String, with options: CompareOptions) -> Bool {
        return range(of: prefix, options: options) != nil
    }

```

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift (hunk 5 above)
Comment (igoriols) [reply WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]: Nice catch. The idea was to have users with a `username` or a `displayName` beginning with the prefix first. I think your dictionary approach suggestion will cover that. I will add a new entry to test this scenario, also.

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift (hunk 6 above)
Comment (igoriols) [reply WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]: That's awesome, I didn't know about `Dictionary(grouping:by:)`. Thanks for sharing! I will apply your suggestion.

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift (hunk 7 above)
Comment (igoriols) [reply WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]: Yes, I forgot to update the docs. Thanks for noticing it.

File: WordPress/WordPressShareExtension/String+Extensions.swift (hunk 4 above)
Comment (AliSoftware) [reply WordPress/WordPressShareExtension/String+Extensions.swift]: I like it 👍

Comment (hassaanelgarem) [reply WordPress/WordPressShareExtension/String+Extensions.swift]: This new way is a bit risky. If someone doesn't pass `.anchored` to this function, it will lose its definition as it will no longer be looking for prefixes. I think we'd have to make the API more strict than this to avoid misuse. Wdyt?

File: WordPress/Classes/Models/UserSuggestion+Comparable.swift [hunk 8]
```diff
@@ -0,0 +1,11 @@
+extension UserSuggestion: Comparable {
+    public static func < (lhs: UserSuggestion, rhs: UserSuggestion) -> Bool {
+        if let leftDisplayName = lhs.displayName, let rightDisplayName = rhs.displayName {
+            return leftDisplayName.localizedCaseInsensitiveCompare(rightDisplayName) == .orderedAscending
+        } else if let leftUsername = lhs.username, let rightUsername = rhs.username {
+            return leftUsername < rightUsername
```
Comment (hassaanelgarem) [inline WordPress/Classes/Models/UserSuggestion+Comparable.swift:6]: Why aren't we using `localizedCaseInsensitiveCompare` here?

File: WordPress/WordPressShareExtension/String+Extensions.swift (hunk 4 above)
Comment (AliSoftware) [reply WordPress/WordPressShareExtension/String+Extensions.swift]: Nice catch and very good point @hassaanelgarem . What about keeping the method name and signature as `hasPrefix(_ prefix: String, with options: CompareOptions)` but make sure the implementation adds `.anchored` to the `options` optionset before using it in the `range(…)` call?

File: WordPress/Classes/Models/UserSuggestion+Comparable.swift (hunk 8 above)
Comment (AliSoftware) [reply WordPress/Classes/Models/UserSuggestion+Comparable.swift:6]: @hassaanelgarem Quoting @igoriols from pbMoDN-3N0-p2#implementation:

> I’m using `localizedCaseInsensitiveCompare` [to compare displayNames], because the user can have the `displayName` in different languages. The comparison is localizable and case-insensitive.
>
> I’m using `<` to compare usernames because they are very restricted, and the lexicographical comparison should work for them.

File: WordPress/WordPressShareExtension/String+Extensions.swift (hunk 4 above)
Comment (igoriols) [reply WordPress/WordPressShareExtension/String+Extensions.swift]: Hey @hassaanelgarem, what if we use a `preconditionFailure` call along with @AliSoftware's suggestion?
```swift
    /// Returns a Boolean value indicating if this String begins with provided prefix.
    /// - Parameter prefix: The provided prefix to verify.
    /// - Parameter options: The provided compare options. It must contain CompareOption.anchored.
    /// - Returns: true if this String begins with the provided prefix following the given compare options, false otherwise.
    func hasPrefix(_ prefix: String, with options: CompareOptions) -> Bool {
        if !options.contains(.anchored) {
            preconditionFailure("options must contain CompareOption.anchored")
        }
        return range(of: prefix, options: options) != nil
    }
```

Comment (AliSoftware) [reply WordPress/WordPressShareExtension/String+Extensions.swift]: `preconditionFailure` not only will only be caught at runtime, but will also put the responsibility to the caller to have to specify an option they shouldn't have to worry about because they shouldn't have to know about the implementation detail of your `hasPrefix` method (ie the fact that it uses `range(of:…)` under the hood — as opposed to some other possible implementation you could have used here, like `self.folding(options: options).hasPrefix(prefix.folding(options: options))` for example, which would have been equally valid without requiring the caller to provide `.anchored` as part of the options at call site.

So TL;DR: it shouldn't be the caller's responsibility to add `.anchored` to the options. The need for `.anchored` is an implementation detail, not a caller requirement.

Comment (igoriols) [reply WordPress/WordPressShareExtension/String+Extensions.swift]: Yes, that makes total sense @AliSoftware. So, let me go back to your original suggestion:
```swift
    func hasPrefix(_ prefix: String, with options: CompareOptions) -> Bool {
        var optionsSet = options
        if !optionsSet.contains(.anchored) {
            optionsSet.update(with: .anchored)
        }
        return range(of: prefix, options: optionsSet) != nil
    }
```
@hassaanelgarem wdyt?

Comment (AliSoftware) [reply WordPress/WordPressShareExtension/String+Extensions.swift]: Since `CompareOptions` is an `OptionSet` (so a set and not an `Array`) I think you can add the `.anchored` flag to the set unconditionally.

```swift
func hasPrefix(_ prefix: String, with options: CompareOptions) -> Bool {
    let fullOptions = options.union([.anchored])
    return range(of: prefix, options: fullOptions) != nil
}
```

Comment (hassaanelgarem) [reply WordPress/WordPressShareExtension/String+Extensions.swift]: @igoriols I agree with @AliSoftware's suggestion 👌 
It's a neat solution and ensures the function is used correctly.

Comment (igoriols) [reply WordPress/WordPressShareExtension/String+Extensions.swift]: Thanks, @hassaanelgarem and @AliSoftware! I applied the changes you suggested.

**Review by hassaanelgarem — APPROVED** [review]
Works as expected 🚀

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift [hunk 9]
```diff
@@ -34,6 +34,8 @@ import CoreData
 
     // MARK: - State
 
+    private static let compareOptions: String.CompareOptions = [.anchored, .caseInsensitive, .diacriticInsensitive]
```
Comment (AliSoftware) [inline WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]: 💄  Nitpick: Not sure it's worth to have this constant as a `private static let` on class level here while the only place where this is used is in the implementation of `sort(userSuggestions:by:)`. 🙃 

To be honest when I read the implementation of your `sort` method below I wondered where the `compareOptions` constant was coming from, until I saw it was a class constant and not just a `let` local to the function, so that was a tad confusing 😅 Maybe it's just me though 😛

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift [hunk 10]
```diff
@@ -123,15 +125,30 @@ import CoreData
...
             return .sites(siteSuggestions)
         }
     }
 
+    /// Sort user suggestions by prefix first, then alphabetically. The collection is sorted first by checking if username or displayName begins with the provided prefix. The remaining items are sorted alphabetically by displayName. The prefix comparison is both case-insensitive and diacritic-insensitive.
+    /// - Parameter userSuggestions: The user suggestions collection to be sorted.
+    /// - Parameter prefix: The prefix to be used when checking the usernames and displayNames.
+    /// - Returns: The sorted user suggestion collection.
+    private static func sort(userSuggestions: [UserSuggestion], by prefix: String) -> [UserSuggestion] {
+        guard !userSuggestions.isEmpty, !prefix.isEmpty else { return userSuggestions }
```
Comment (AliSoftware) [inline WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift:143]: 💄  Nitpick: we could just define `let compareOptions: String.CompareOptions = [.anchored, .caseInsensitive, .diacriticInsensitive]` here instead of having it as `private static let` at the top of the file and having to scroll to find it (see also previous comment)

File: WordPress/WordPressShareExtension/String+Extensions.swift [hunk 11]
```diff
@@ -88,6 +88,15 @@ extension String {
         return self + returnURLString
     }
 
+    /// Returns a Boolean value indicating if this String begins with provided prefix.
+    /// - Parameter prefix: The provided prefix to verify.
+    /// - Parameter options: The provided compare options.
```
Comment (AliSoftware) [inline WordPress/WordPressShareExtension/String+Extensions.swift]: 💄  Not sure if this is some documentation convention/habit but I found it quite original that you use "provided" in the parameters documentation… the parameters are always something that the caller is provided, so felt a bit unusual to have that wording here 😄 

What do you think of this instead?
```suggestion
    /// Returns a Boolean value indicating if this String begins with the provided prefix.
    /// - Parameter prefix: The prefix to check for.
    /// - Parameter options: The string comparison options to use when checking for the prefix.
```

File: WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift [hunk 12]
```diff
@@ -190,7 +190,7 @@ class SuggestionsListViewModelTests: CoreDataTestCase {
 
     // MARK: - Helpers
 
-    private func suggestionsList(fromProminentIds prominentIds: [Int], regularIds: [Int], in userSuggestions: [UserSuggestion]) -> [SuggestionsListSection] {
+    private func suggestionsList(fromProminentIds prominentIds: [Int], regularIds: [Int]) -> [SuggestionsListSection] {
```
Comment (AliSoftware) [inline WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift:221]: Nice cleanup! 👍 👏 

_PS: 🤔  Completely out of curiosity: was there a warning that told you that `in userSuggestions: [UserSuggestion]` was unused? Would be pretty nice of the compiler if it was detecting such a case, but I'm not sure if today's Swift Compiler warns us about such cases, so just curious 🙂_

File: WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift [hunk 13]
```diff
@@ -119,7 +119,7 @@ class SuggestionsListViewModelTests: CoreDataTestCase {
     func testSearchSuggestionsWithPartialMatch() throws {
         // Given
         let word = "@ca"
```
Comment (AliSoftware) [inline WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift:149]: 💡  I notice that you don't have tests that contains diacritics or uppercase letters as prefix. Which means we don't cover the cases of things like `@Ca`, or ensure that e.g. `@Caa` would match user with display name `"Caárla Diaz"`.

Could be nice to add some coverage for that, especially given you were thoughtful in the implementation to implement case-insensitive and diacritics-insensitive comparison for display Name (but not for user name), so would be nice to test for it to ensure there won't be an accidental regression for that aspect in the future.

**Review by AliSoftware — COMMENTED** [review]
Overall good, nice job!

Left a couple of nitpick suggestions (up to you to apply or not), and also a suggestion for additional test cases. But other than that this LGTM 👍

File: WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift (hunk 12 above)
Comment (igoriols) [reply WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift:221]: Unfortunately, there was no warning. I agree with you that the compiler could help us with that. Perhaps in a future version 😄 .

File: WordPress/WordPressShareExtension/String+Extensions.swift (hunk 11 above)
Comment (igoriols) [reply WordPress/WordPressShareExtension/String+Extensions.swift]: I checked other documentation around the project, and `provided` was used a few times. As a non-English native speaker, I agree with you that it can be improved. I'm going to apply your suggestion. Thank you!

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift (hunk 9 above)
Comment (igoriols) [reply WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift]: I used this approach not to have always to create `compareOptions`. I will move to the `sort` method to clarify.

File: WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift (hunk 10 above)
Comment (igoriols) [reply WordPress/Classes/ViewRelated/SuggestionsListViewModel.swift:143]: It makes sense. I will move to the `sort` method to clarify.

File: WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift (hunk 13 above)
Comment (igoriols) [reply WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift:149]: Nice catch. I added more tests to cover more scenarios. Thank you!

File: WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift [hunk 14]
```diff
@@ -115,11 +115,39 @@ class SuggestionsListViewModelTests: CoreDataTestCase {
         XCTAssertTrue(result)
     }
 
+    /// Tests that the seach result has one match with case-insensitive and diacritic-insensitive
```
Comment (AliSoftware) [inline WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift]: Just noticed this small typo (and you got the same in other test cases below)
```suggestion
    /// Tests that the search result has one match with case-insensitive and diacritic-insensitive
```

**Comment by AliSoftware** [comment]
Just thought of some edge cases (though that one should probably be for a separate PR rather than this one): how does this whole thing behaves with RTL and non-Latin languages?

 - I'd assume that thanks to the `localizedCaseInsensitiveCompare` the comparison will be locale-aware, so we should be good on that front with using the sort order according to the user's locale 👍
 - But is that also true for our `hasPrefix`, especially for RTL text? Or using Latin transcription prefix of simplified Chinese characters? 🤔 Not sure that `.caseInsensitive, .diacriticsInsensitive` covers locale insensitive or awareness 🤔
 - And how does both the sorting and the filtering fares when the list of suggested users contains a mix of users with Latin displayName and other users with eg Japanese or Chinese characters in their display name? I _think_ it should turn out OK (because `Foundation` is smart enough to deal with those on our behalf in the `localizedCaseInsensitiveCompare` and `range` implementations, but would be nice to confirm 😛
 - And is the logic of "The user types `@` then `C` then `a` in the  textField and expect the filtering to be by prefix `Ca` (and not `aC` or `\u{200F}Ca`) (U200F being the RTL marker in Unicode, though I doubt it would be present in the String, since the Bidi algorithm would take care of switching the text direction between @ and the text as needed automatically… but worth checking)
 - If the user types `@Cá` with the acute accent explicitly typed, should we list suggestions actually having the accent first, then the ones matching in diacritics-insensitive way next, then the rest? This would probably be taking it a bit too far and not worth the effort tho 😅 

To be clear, all of those cases/open questions are (1) likely out of scope for this PR (2) might not all be worth actually fixing or even investigating, even in a separate PR — as I'm probably overthinking some of those cases. Still, I just thought I'd braindump those here to raise the question and see if it's worth checking/ investigating those cases or not. 🙃

**Review by AliSoftware — APPROVED** [review]
Apart from the small 💄  [typo in the test comments](https://github.com/wordpress-mobile/WordPress-iOS/pull/19218#discussion_r960088313), this looks ready to merge to me now. Nice work 👍

File: WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift (hunk 14 above)
Comment (igoriols) [reply WordPress/WordPressTest/Mention/SuggestionsListViewModelTests.swift]: Fixed. Thank you!

**Comment by hassaanelgarem** [comment]
@igoriols I updated the milestone to 20.8 since we won't catch the code freeze. Please update the release note accordingly 🙏

**Comment by mokagio** [comment]
@igoriols just wanted to let you know that WordPress 20.8, the first version with your changes from this PR, just started rolling out to users.

Thank you so much for your contribution! 🎉
